### ISSUE TITLE
Arrow-up and down not working in "Find References" tree when NVDA is enabled

### ISSUE THREAD

**Issue by WiproPSAccTest** [Microsoft/vscode#17245]
VS Code Version: 1.6.1
OS Version:1607
NVDA version:-2016.2.1

**Steps to Reproduce:**
Prerequisite: Node.js and git repository need to be installed.
1.Open Visual studio code editor.
2.Open any file and using "Ctrl+1" key  navigate to the  first vs code editor.
3.Select any code and using "Shift+F12", key apply "Find All References" to the code  present in the editor.

**Expected result:**
NVDA should recite all the available reference for the selected code.
EX:-When we apply "Shift+F12" for "Find All References" then NVDA should recite all the reference available to that particular code.

**Actual Result:**
NVDA is not announcing the available references for the selected code in the editor.

**User Impact:**
Screen Reader user doesn't get to know about results of find all references as it's not getting announced.

Attachments:
![nvda_speechviewerinformation_vscodeeditor_findallreference](https://cloud.githubusercontent.com/assets/22956130/21213559/37a09996-c2bb-11e6-9c9d-abafedca835d.jpg)

**Comment by jrieken** [comment]
@bpasero I am using the `accessibilityProvider`-API you have added and I do see that the items have an `aria-label` property but nothing is being read. Is there something else in the API to enable accessibility in the tree?

**Comment by bpasero** [comment]
@jrieken no, the provider should take care of everything. that is, all it does is doing the ARIA wiring up. If something does not work there, it must be an issue with something else (e.g. maybe focus is not inside the tree?).

I can look into it.

**Comment by jrieken** [comment]
Thanks. Provider is implemented here: https://github.com/Microsoft/vscode/blob/master/src/vs/editor/contrib/referenceSearch/browser/referencesWidget.ts#L409

**Comment by bpasero** [comment]
Looks ok to me, the tree (when it has focus) sets the `aria-activedescendent` property to the ID of the element that has focus and that element has an `aria-label`. We do the same thing in other trees and it works. We do this because only the root node of the tree has a `tabindex` but we want screen readers to read the text of an item within (the selected tree item).

![image](https://cloud.githubusercontent.com/assets/900690/25344494/9548356e-2912-11e7-9896-72c20cb7f9c1.png)

I can only speculate that the fact that the tree is embedded within the editor is causing issues. E.g. if a parent has role="presentation" set, the screen reader might ignore everything within.

Another thing that I noticed is that, as soon as I have NVDA running, I can no longer use the arrow keys to navigate up and down in the references. Might be related.

**Comment by jrieken** [comment]
Hm... @alexandrudima I guess `role="presentation"` is there for a reason. I might be able to help myself by using explicit `alert`s when the tree changes or by ignoring the tree and simply reading all results when they are computed. Both not ideal...

**Comment by jrieken** [comment]
I have now pushed a manual-alert-solution (which is what we also use for the suggest widget)

**Comment by bpasero** [comment]
Ok good, still weird that arrow-keys do not work once NVDA is enabled. I suspect some general f-up in that area is causing all kinds of weird accessibility issues.

**Comment by chrmarti** [comment]
Hard to interpret:
```
Found 6 symbols in c:\devel\vscode-regex\src\extension.ts (occurred again) 
clickable
clickable
clickable
Find  edit  unavailable  Find
 
clickable
Match Case (Alt+C)
check box  unavailable  not checked  Match Case (Alt+C)
Match Whole Word (Alt+W)
check box  unavailable  not checked  Match Whole Word (Alt+W)
Use Regular Expression (Alt+R)
check box  unavailable  not checked  Use Regular Expression (Alt+R)
clickable
No Results
Previous match (Shift+F3)
button  unavailable  Previous match (Shift+F3)
Next match (F3)
button  unavailable  Next match (F3)
clickable
Find in selection
 
check box  unavailable  not checked
Close (Escape)
```

![screen shot 2017-04-27 at 4 20 51 pm](https://cloud.githubusercontent.com/assets/9205389/25508232/881ba240-2b65-11e7-8f07-684b7f8c1f6a.png)

This is when invoking the command, no navigation through the results. @jrieken is this expected?

**Comment by jrieken** [comment]
I think the problem is that with NVDA the arrow keys aren't working and therefore don't read the individual references while going up and down. @alexandrudima Unsure why it reads the editor search box?

**Comment by dbaeumer** [comment]
I hear the same as @chrmarti. Should we reopen this ?

**Comment by jrieken** [comment]
Unsure, I hear something on Mac using VoiceOver. Can someone try a different windows tool?

**Comment by dbaeumer** [comment]
I tested Narrator and that is even worse. All it reads is arrow up, down when pressed in the references list.

**Comment by jrieken** [comment]
wtf - I'll reopen tho I have no clue how to fix this...

**Comment by bpasero** [comment]
I am getting good things reported to NVDA when clicking on elements in the tree:

![image](https://cloud.githubusercontent.com/assets/900690/25608484/4b04a46a-2f1b-11e7-95e4-ac435ea5605f.png)

Updating the summary to reflect the actual issue.

**Comment by bpasero** [comment]
It is weird, I can arrow up/down enough to eventually get the editor contents announced in the screen reader. I also loose keyboard focus from the find references tree eventually.

**Comment by jrieken** [comment]
/cc @bramd @jcsteh because you have shown NVDA expertise in the past. The challenge is to figure out why arrow keys (up/down) don't work when invoking reference search. The speech log below indicates to me that instead of selecting and reading tree nodes, parts of the status bar are read. At least, that's where you find "Tweet Feedback" unsure where "clickable", "main landmark", or "content info landmark" come from?

```

        const updatedSetting = vscode.workspace.getConfiguration('extension').get('setting');

blank
Found 4 symbols in 2 files 
main landmark 
clickable
clickable
clickable
content info landmark 
Tweet Feedback
clickable
 
symbol in \\localHOST\c$\Users\jrieken\code\virtual\src\extension.ts on line 19 at column 39 
main landmark 
clickable
clickable
clickable
```

**Comment by jcsteh** [comment]
You're in NVDA browse mode, which is used to give users a document like view of web content. So, NVDA intercepts the arrow keys to move by line, word, character, etc.

You could make NVDA switch to focus mode by focusing something with an interactive role. The fact that this isn't happening already suggests you're focusing something which doesn't have an interactive role. What is getting focus here?

An alert to output this kind of content is almost always a bad thing. If it's a tree, you should be using (and focusing) an ARIA tree. It seems you tried to do that, but it didn't work for some reason:

[**@bpasero**](https://github.com/bpasero) commented on [25 Apr 2017, 01:26 GMT+10](https://github.com/Microsoft/vscode/issues/17245#issuecomment-296705927 "2017-04-24T15:26:49Z - Replied by Github Reply Comments"):
> I can only speculate that the fact that the tree is embedded within the editor is causing issues.

When you say "embedded within the editor", what do you mean? Is it inside a contentEditable or do you mean something else?

> E.g. if a parent has `role="presentation"` set, the screen reader might ignore everything within.

It shouldn't do. I can tab to this tree just fine in Chrome:

`data:text/html,<div role="presentation"><div role="tree" tabindex="0" aria-activedescendant="item"><div role="treeitem" id="item">foo</div></div></div>`
Btw, in a tree, the expected result isn't to report all items. It would only report the focused item.

**Comment by jrieken** [comment]
@jcsteh Thanks for the explanation. That helped my fix this issue which was actually caused by us having a code path setting in-editor-views to be aria-hidden (added in https://github.com/Microsoft/vscode/commit/ace4741833400f041bc8f4b08a06465cb731af18)
Also, I have removed all aria-alert-calls in the symbols world as things now work much better. Give it a try, very much open for more feedback. Happy Coding!
